rfc/0032-field-level-encryption.md

```diff
@@ -283,10 +283,11 @@ Where:
 | ------------- |:-------------:|
 |PREFIX | Prefix for temporary field that holds content of encrypted field. This should be configurable so that naming conflicts can be easily resolved at the application level. The default is “crypt”.|
 |FIELDNAME |The original name of the field that contained the data to be encrypted |
+|BASE64() | Standard Base64 encoding. The results MUST include padding if applicable. Padding is required because the encoded results are included in signature calculation, which must be consistent across implementations.|
 |“kid”| The “key-identifier” for resolving the key used to decrypt/encrypt from the KeyStore. See section on Key Management above.|
 |“alg” |The algorithm used to encrypt/decrypt. |
 |“ciphertext” | A Base64 encoded string that is the value from the field that has been encrypted.|
-|“sig”|Optional, required for AES. The HMAC signature of the following fields concatenating and then base64 encoding them: “alg”, “iv”, and “ciphertext”. Order is important and must be respected across all implementations |
+|“sig”|Optional, required for AES. The HMAC signature, calculated as described above. |
```

Comment:
This would be a breaking change I believe between SDK2 and SDK3, would it not?

Reply:
The 'as described above' is reflecting that there is a description of this field directly above the table already.  We should make sure that the algorithm reflected above actually mirrors the implementation that ended up in the RFCs though.

Reply:
For reference, the Node.js implementation is:
BASE64(HMAC_SHA256(hmacKey, kid + alg + iv + ciphertext))
Hopefully this is reflected by other SDKs considering it wouldn't make a lot of sense to be BASE64 encoding the contents of an HMAC signature, they should just be concatenated.

Reply:
I agree it makes no sense to Base64 encode the signature components. Unfortunately, the Java implementation Base64 encodes the IV and ciphertext when calculating the signature. It's probably doing this because the RFC says:

>The value of sig can be computed as follows:
>
>```
>sig = BASE64(HMAC256(kid + alg + BASE64(iv) + BASE64(ciphertext))))
>```

For what it's worth, the example signatures in [this section of the RFC](https://github.com/couchbaselabs/sdk-rfcs/blob/master/rfc/0032-field-level-encryption.md#ciphers) can only be matched if the IV and ciphertext are Base64 encoded when calculating the signature.

Reply:
I think the way forward is probably to update the RFC to match what .Net and Node.js are doing. The Java implementation can be updated to generate the correct signature but accept both versions.

Reply:
@dnault - that is a bug in the RFC, it should be `sig = HMAC256(BASE64(kid + alg + iv + ciphertext))`